x/wasm/keeper/keeper_test.go

```diff
@@ -588,28 +609,64 @@ func TestExecute(t *testing.T) {
 	require.NoError(t, err)
 	require.Equal(t, "cosmos14hj2tavq8fpesdwxxcu44rty3hh90vhujrvcmstl4zr3txmfvw9s4hmalr", addr.String())
 
+	size, err := GetSize(ctx, addr, keepers.WasmKeeper.storeKey)
+	require.Nil(t, err)
+	assert.Equal(t, uint64(185), size)
+
+	rentInfo, err := GetRentInfo(ctx, addr, keepers.WasmKeeper.storeKey)
+	require.Nil(t, err)
+	assert.Equal(t, int64(0), rentInfo.Balance)
+	assert.Equal(t, uint64(ctx.BlockHeight()), rentInfo.LastChargedBlock)
+
 	// ensure bob doesn't exist
 	bobAcct := accKeeper.GetAccount(ctx, bob)
 	require.Nil(t, bobAcct)
 
 	// ensure funder has reduced balance
 	creatorAcct := accKeeper.GetAccount(ctx, creator)
 	require.NotNil(t, creatorAcct)
-	// we started at 2*deposit, should have spent one above
-	assert.Equal(t, deposit, bankKeeper.GetAllBalances(ctx, creatorAcct.GetAddress()))
+	// we started at 3*deposit, should have spent one above
+	assert.Equal(t, deposit.Add(deposit...), bankKeeper.GetAllBalances(ctx, creatorAcct.GetAddress()))
 
 	// ensure contract has updated balance
 	contractAcct := accKeeper.GetAccount(ctx, addr)
 	require.NotNil(t, contractAcct)
 	assert.Equal(t, deposit, bankKeeper.GetAllBalances(ctx, contractAcct.GetAddress()))
 
+	// bump block height
+	ctx = ctx.WithBlockHeight(ctx.BlockHeight() + 2)
+
+	// insufficient rent
+	res, err := keepers.ContractKeeper.Execute(ctx.WithEventManager(sdk.NewEventManager()), addr, fred, []byte(`{"release":{}}`), topUp)
+	require.Error(t, err)
+	require.True(t, errors.Is(err, types.ErrInsufficientRent))
+
+	// deposit rent
+	err = keepers.WasmKeeper.DepositRent(ctx, addr, creator, 100000)
+	require.Nil(t, err)
+	rentInfo, err = GetRentInfo(ctx, addr, keepers.WasmKeeper.storeKey)
+	require.Nil(t, err)
+	assert.Equal(t, int64(99996), rentInfo.Balance)
+
 	// unauthorized - trialCtx so we don't change state
 	trialCtx := ctx.WithMultiStore(ctx.MultiStore().CacheWrap(keepers.WasmKeeper.storeKey).(sdk.MultiStore))
-	res, err := keepers.ContractKeeper.Execute(trialCtx, addr, creator, []byte(`{"release":{}}`), nil)
+	res, err = keepers.ContractKeeper.Execute(trialCtx, addr, creator, []byte(`{"release":{}}`), nil)
 	require.Error(t, err)
 	require.True(t, errors.Is(err, types.ErrExecuteFailed))
 	require.Equal(t, "Unauthorized: execute wasm contract failed", err.Error())
 
+	size, err = GetSize(ctx, addr, keepers.WasmKeeper.storeKey)
+	require.Nil(t, err)
+	assert.Equal(t, uint64(185), size)
+
+	rentInfo, err = GetRentInfo(ctx, addr, keepers.WasmKeeper.storeKey)
+	require.Nil(t, err)
+	assert.Equal(t, int64(99996), rentInfo.Balance)
+	assert.Equal(t, uint64(ctx.BlockHeight()), rentInfo.LastChargedBlock)
+
+	// bump block height
+	ctx = ctx.WithBlockHeight(ctx.BlockHeight() + 1)
```

Comment:
can we add a test case for two txs in the same block that modify state size each?

Reply:
yeah good call